File: app/src/main/java/com/dragonpav/dengine/StaticModel.java
package com.dragonpav.dengine;

import android.opengl.GLES30;
import android.util.Log;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.FloatBuffer;
import java.nio.IntBuffer;
import java.util.ArrayList;

import glm_.mat4x4.Mat4;

public class StaticModel {
    public Values.Matrix4 model;
    public FloatBuffer vertexBuffer;
    public IntBuffer indexBuffer;
    public FloatBuffer texBuffer;
    public FloatBuffer normalBuffer;
    public Texture texture;
    public int vertBufSize, indexBufSize, texBufSize, normalBufSize;
    public ObjectBuffer b;
    Camera cam;
    RenderUtils renderUtils;
    public StaticModel(Camera camera, RenderUtils rend, ArrayList<Object3D> objects) {
        cam = camera;
        renderUtils = rend;
        texture = objects.get(0).texture;
        boolean usesLighting = rend.lighting != null;
        int indexOffset = 0;
        for (Object3D obj : objects) {
            vertBufSize += obj.vertexBufferSize;
            indexBufSize += obj.indexBufferSize;
            texBufSize += obj.texBufferSize;
            normalBufSize += obj.normalBufferSize;
        }
        vertexBuffer = ByteBuffer.allocateDirect(vertBufSize)
                .order(ByteOrder.nativeOrder())
                .asFloatBuffer();
        indexBuffer = ByteBuffer.allocateDirect(indexBufSize)
                .order(ByteOrder.nativeOrder())
                .asIntBuffer();
        texBuffer = ByteBuffer.allocateDirect(texBufSize)
                .order(ByteOrder.nativeOrder())
                .asFloatBuffer();
        if (usesLighting) {
            normalBuffer = ByteBuffer.allocateDirect(normalBufSize)
                    .order(ByteOrder.nativeOrder())
                    .asFloatBuffer();
        }
        for (Object3D obj : objects) {
            float[] verts = new float[obj.vertexBuffer.capacity()];
            obj.vertexBuffer.get(verts);
            vertexBuffer.put(verts);
            obj.vertexBuffer.position(0);
            for (int i = 0; i < obj.indexBuffer.capacity(); i++) {
                indexBuffer.put(obj.indexBuffer.get(i) + indexOffset);
            }
            float[] tex = new float[obj.texBuffer.capacity()];
            obj.texBuffer.get(tex);
            texBuffer.put(tex);
            obj.texBuffer.position(0);
            if (usesLighting) {
                float[] norms = new float[obj.normalBuffer.capacity()];
                obj.normalBuffer.get(norms);
                normalBuffer.put(norms);
                obj.normalBuffer.position(0);
            }
            indexOffset += obj.vertexBuffer.capacity() / 3;
        }
        vertexBuffer.position(0);
        indexBuffer.position(0);
        texBuffer.position(0);
        if (usesLighting) normalBuffer.position(0);
        model = new Values.Matrix4(GLES30.glGetUniformLocation(rend.p.programId, "model"));
        b = renderUtils.createStaticVAO(vertexBuffer, indexBuffer, texBuffer, normalBuffer, vertBufSize, indexBufSize, texBufSize, normalBufSize);
    }
    public void begin() {
        model.values = new Mat4(1.0f);
        GLES30.glBindVertexArray(b.VAO);
    }
    public void render() {
        model.uniform();
        GLES30.glActiveTexture(texture.glType);
        GLES30.glBindTexture(GLES30.GL_TEXTURE_2D, texture.textureId);
        if (cam.autoUpdate) {
            cam.update();
        }
        if (cam.rendUtils.lighting != null) {
            cam.rendUtils.lighting.bind();
        }
        if (texture.texSampleNum == -1) {
            Log.e("DEngine", "texture.texSampleNum is -1");
            return;
        }
        else {
            GLES30.glUniform1i(texture.texSampleNum, texture.glType % GLES30.GL_TEXTURE0);
        }
        GLES30.glDrawElements(GLES30.GL_TRIANGLES, indexBufSize / 4, GLES30.GL_UNSIGNED_INT, 0);
    }
    public void end() {
        GLES30.glBindTexture(GLES30.GL_TEXTURE_2D, 0);
        GLES30.glBindVertexArray(0);
    }
    public void dispose() {
        end();
        GLES30.glDeleteVertexArrays(1, new int[] {b.VAO}, 0);
        ArrayList<Integer> buffersToDelete = new ArrayList<>();
        buffersToDelete.add(b.VBO);
        buffersToDelete.add(b.VBOtexture);
        buffersToDelete.add(b.EBO);
        if (cam.rendUtils.lighting != null) buffersToDelete.add(b.VBOnormal);
        int[] tmp = buffersToDelete.stream().mapToInt(Integer::intValue).toArray();
        GLES30.glDeleteBuffers(buffersToDelete.size(), tmp, 0);
        b = new ObjectBuffer();
    }
}

File: src/StaticModel.cpp
#include "StaticModel.h"
StaticModel::StaticModel(Camera* camera, RenderUtils* rend, std::vector<Object3D>& objs) {
	if (camera == nullptr) {
		throw std::runtime_error("Camera is null");
	}
	if (rend == nullptr) {
		throw std::runtime_error("RenderUtils is null");
	}
	if (objs.empty()) {
		throw std::runtime_error("Objects is empty");
	}
	this->cam = camera;
	renderUtils = rend;
	texture = objs[0].getTexture();
	bool usesLightning = objs[0].getCamera()->getRenderUtils()->getLighting() != nullptr;

	size_t totalVertices = 0;
	size_t totalIndices = 0;
	size_t totalTexCoords = 0;
	size_t totalNormals = 0;

	for (auto& obj : objs) {
		totalVertices += obj.vertexBuffer.size();
		totalIndices += obj.indexBuffer.size();
		totalTexCoords += obj.texBuffer.size();
		totalNormals += obj.normalBuffer.size();
	}

	vertexBuffer.reserve(totalVertices);
	indexBuffer.reserve(totalIndices);
	texBuffer.reserve(totalTexCoords);
	if (usesLightning) {
		normalBuffer.reserve(totalNormals);
	}

	GLuint indexOffset = 0;
	for (const auto& obj : objs) {
		vertexBuffer.insert(vertexBuffer.end(), obj.vertexBuffer.begin(), obj.vertexBuffer.end());
		for (auto index : obj.indexBuffer) {
			indexBuffer.push_back(index + indexOffset);
		}
		texBuffer.insert(texBuffer.end(), obj.texBuffer.begin(), obj.texBuffer.end());
		if (usesLightning) {
			normalBuffer.insert(normalBuffer.end(), obj.normalBuffer.begin(), obj.normalBuffer.end());
		}
		indexOffset += (GLuint) obj.vertexBuffer.size() / 3;
	}

	vertBufSize = (GLuint) vertexBuffer.size() * sizeof(GLfloat);
	indexBufSize = (GLuint) indexBuffer.size() * sizeof(GLuint);
	texBufSize = (GLuint) texBuffer.size() * sizeof(GLfloat);
	normalBufSize = (GLuint) normalBuffer.size() * sizeof(GLfloat);
	model = Matrix4(glGetUniformLocation(rend->getProgram()->getId(), "model"));
	b = renderUtils->createStaticVAO(vertexBuffer.data(), indexBuffer.data(), texBuffer.data(), normalBuffer.data(), vertBufSize, indexBufSize, texBufSize, normalBufSize);
}
StaticModel::~StaticModel() {
	dispose();
}
void StaticModel::setTexture(Texture* tex) {
	texture = tex;
}
void StaticModel::setCamera(Camera* camera) {
	cam = camera;
}
Texture* StaticModel::getTexture() {
	return texture;
}
Camera* StaticModel::getCamera() {
	return cam;
}
Matrix4& StaticModel::getModelMatrix() {
	return model;
}
void StaticModel::begin() {
	model.values = glm::mat4(1.0f);
	glBindVertexArray(b.VAO);
}
void StaticModel::render(Program* currentProgram, bool isShadowPass) {
	if (currentProgram == nullptr) return;
	GLint currentModelLoc = glGetUniformLocation(currentProgram->getId(), "model");
	if (currentModelLoc != -1) {
		model.uniformL = currentModelLoc;
		model.uniform();
	}
	if (isShadowPass) {
		glDrawElements(GL_TRIANGLES, static_cast<GLsizei>(indexBuffer.size()), GL_UNSIGNED_INT, 0);
		return;
	}
	glActiveTexture(texture->getGLType());
	glBindTexture(GL_TEXTURE_2D, texture->getTextureId());
	if (cam->isAutoUpdate()) {
		cam->update();
	}
	if (cam->getRenderUtils()->getLighting() != nullptr) {
		cam->getRenderUtils()->getLighting()->bind();
	}
	if (texture->getTextureSample() == -1) {
		std::cerr << "texture.texSampleNum is -1" << std::endl;
		return;
	}
	else {
		glUniform1i(texture->getTextureSample(), texture->getGLType() % GL_TEXTURE0);
	}
	glDrawElements(GL_TRIANGLES, static_cast<GLsizei>(indexBuffer.size()), GL_UNSIGNED_INT, 0);
}
void StaticModel::dispose() {
	glBindTexture(GL_TEXTURE_2D, 0);
	glBindVertexArray(0);
	glDeleteVertexArrays(1, &b.VAO);
	std::vector<GLuint> buffersToDelete;
	buffersToDelete.push_back(b.VBO);
	buffersToDelete.push_back(b.VBOtexture);
	buffersToDelete.push_back(b.EBO);
	if (cam->getRenderUtils()->getLighting() != nullptr) buffersToDelete.push_back(b.VBOnormal);
	glDeleteBuffers(static_cast<GLsizei>(buffersToDelete.size()), buffersToDelete.data());
	b = ObjectBuffer();
}